app/scripts/athena_partition_archive.py

```diff
@@ -286,6 +289,176 @@ def execute_athena(sql, database_name):
 
     return response
 
+def partition(database_name, table_name, s3_location, retention):
+    """
+    Gets a list of partitions from Athena, then removes partitions based on the retention period.
+
+    Args:
+        database_name  : the schema name in Athena
+        table_name     : the table name in Athena
+        s3_location    : the S3 location of the data in the schema
+        retention      : retention period / days to keep
+
+    Returns:
+        None
+    """
+
+    try:
+        LOGGER.info('Processing %s.%s, removing partitions older than %s.', database_name, table_name, retention)
+
+        sql = "show partitions " + database_name + "." + table_name
+
+        response = execute_athena(sql, database_name)
+        execution_id = response['QueryExecution']['QueryExecutionId']
+
+        query_results = ATHENA.get_query_results(
+            QueryExecutionId=execution_id)
+
+        partition_list = []
+
+        for row in query_results['ResultSet']['Rows']:
+            path_name = row['Data'][0]['VarCharValue']
+            try:
+                match = PATTERN.search(path_name).group(0)
+                if match <= str(retention):
+                    if path_name != 'path_name':
+                        partition_list.append("""path_name={0}""".format(path_name))
+            except:
+                LOGGER.info("No match found.")
+                break
+
+        for item in partition_list:
+            item_quoted = item[:10] + "'" + item[10:] + "'"
+            item_stripped = item.split('=')[1]
+
+            drop_partition_sql = ("ALTER TABLE " + database_name + "." + table_name + \
+                                 " DROP PARTITION (" + item_quoted + ");")
+            add_partition_sql = ("ALTER TABLE " + database_name + "." + table_name + \
```

Comment:
Suggest you "add partition" before you "drop partition" so that if the S3 location does not exist the add will fail and prevent the drop.

Reply:
Swapped add and drop partition so adding of a partition is done first which will also check for the S3 location.